### DISCUSSION TITLE
Additionnal categories for extensions?

### DISCUSSION THREAD

**Discussion by D8H** [4ian/GDevelop#3813] [Enhancement ideas (and their technical discussions)]
It seems that extensions can be organized in 7 categories:
* Movement and positioning
* Camera
* Game mechanics
* Animations and effects
* UI
* Inputs
* Math and technical tools
* Network

But it doesn't really fit the existing categories:
* General (this one shouldn't be suggested as a category but a default/no value)
* Ads
* Audio
* Advance
* Device
* Input
* Network
* Third-party

It could be interesting use this metadata to let users scan the list by category when they don't know the term they should search.

Could some categories be added?

# Movement and positioning

### Animated Back and Forth Movement

Make the object go on the left, then when some distance is reached, flip and go back to the right. Make sure that your object has two animations called “GoLeft” and “TurnLeft”.

### Linear Movement

Move the object linearly, according to the speed configured on X and Y axis. Useful for simple enemies, bullets or objects following a straight line on the screen.

### Linked Objects Tools

Conditions to use Linked Objects as a graph and a path finding movement behavior.

### Rectangular Movement

Allow an object to move in a rectangular pattern.

### Sine (or ellipsis) Movement

Allow an object to move smoothly on the X and/or Y axis following a sine wave, or an ellipsis.

### Timed Back and Forth Movement

This behavior moves objects back and forth for a chosen time or distance, vertically or horizontally.

### Turret movement

A turret movement with customizable speed, acceleration and stop angles.

### Bounce (using forces)

Provides an action to make the object bounce from another object it just touched. Add a permanent force to the object and, when in collision with another one, use the action to make it bounce realistically.

### Advanced jump ("Coyote time")

Platformer character jump enhancements like coyote time (allows to jump shortly after leaving a platform).

### Cancellable draggable object

Allow to cancel the drag of an object (having the Draggable behavior) and return it smoothly to its previous position.

### Boomerang

Throw an object that returns to the thrower like a boomerang.

### Draggable (for physics objects)

Drag a physics object with the mouse (or touch).

### Fire bullets

Allow the object to fire bullets, with customizable speed, angle and fire rate.

### Screen Wrap

Teleport objects leaving one side of the screen so that they immediately reappear on the opposite side, maintaining speed and trajectory.

### Shake Object (position, angle, scale)

Shake an object, using one or more ways to shake (position, angle, scale).

### Snap objects to a virtual grid

Action to snap objects to a virtual grid during the game.

### Hexagonal grid

Action to snap objects to a virtual hexagonal grid during the game.

### Explosion force

Simulate an explosion with physics forces on target objects.

### Homing projectile

Make a projectile object move towards a target object.

### Checkpoints

A position checkpoint for instances.

### Stay On Screen

Force the object to stay visible on the screen by setting back its position inside the viewport of the camera.

### Stick objects to others

Behavior to stick objects to other objects - so that they follow the position and rotation of the object they are stuck to.

### Make object travel to random positions

Make object travel to random positions (with the pathfinding behavior).

### Values of multiple objects

Values of picked object instances (including position, force, and angle).

### Make objects orbit around a center object

Make objects orbit around a center object. You can customize the number of objects, speed, etc…

# Camera

### Camera Shake

Shake the camera on the specified layer using one or more methods of shaking (position, angle, zoom).

### Camera Zoom

Allows to zoom camera on a layer with a speed (factor per second).

### Copy camera settings

Copy the camera settings of a layer and apply them to another layer.

### Drag camera with the mouse (or touchscreen)

Move a camera by dragging the mouse (or touchscreen).

### Follow multiple objects with camera

Change the zoom and position of the camera to keep all instances of an object (or object group) on the screen.

### Parallax for Tiled Sprite

Behaviors to animate Tiled Sprite objects in the background, following the camera with a parallax effect.

### Room-based camera movement

Move and zoom camera to the room object that contains the trigger object (usually the player).

### Smooth Camera

Smoothly scroll to follow an object.

### Object "Is On Screen" Detection

This adds a condition to detect if an object is on screen based off its current layer.

### Idle object tracker

Check if an object has not moved (with some, customizable, tolerance) for a certain duration (1 second by default).

# Game mechanics

### Object spawner

Spawn (create) objects periodically.

### Object Stack

An ordered list of objects and a shuffle action.

### RTS-like unit selection

Allow player to select units by clicking on them or dragging a selection box.

### Rectangular flood fill

Create fill objects that cover the rectangular area of target objects.

### Repeat every X seconds

This allows to trigger an event every X seconds.

### Dungeon generator

Create a procedurally generated dungeon. Includes two layouts: “rooms and hallways” and “dungeon cave”.

### Health (life points and damages for objects)

Add actions and conditions to manage the life of the object. Object can get damages, lose some health and you can check with a condition if it's dead.

### Noise generator

Expressions for generating Simplex noise values. Useful for procedural generation.

### Create multiple copies of an object

Select the object, choose the number of rows and columns, the spacing between rows and columns, the top left starting point, the layer to create the objects on, and the z-order of the objects.

# Animations and effects

### Auto typing animation for texts ("typewriter" effect)

Animate the text to simulate it being written character by character (also called “typewriter” effect), with a customizable time between each character. Useful for dialogue scenes or visual novels.

### Depth effect

Change scale based on Y position to simulate depth of field.

### Face Forward

Face object towards the direction of movement.

### Flash (blink)

Make the object flash (blink) for a period of time, so that it is alternately visible and invisible. Trigger the effect by using the Flash action.

### Flash layer

Make a layer visible for a specified duration, and then hide the layer.

### Flash and transition painter

Behavior for shape painter allows you to paint a color all over the screen for period of time with an effect (useful for simulate flash and transition effect).

### 3D Flip

Flip sprites with a 3D effect.

### YSort

Create an illusion of depth by setting the Z-order based on the Y position of the object. Useful for isometric games, 2D games with a “Top-Down” view, RPG…

### Animate Shadow Clones

Create and animate shadow clones that follow the path of a primary object.

### Object Masking

Use a sprite or a shape painter to mask another object.

### Sprite Sheet Animations

Animate a tiled sprite from a sprite sheet.

### Reflection

Add a reflection to a selected object.

### Sprite Snapshot

Renders an object, layer or scene and puts the resulting image into a sprite.

### Marching Squares (experimental)

Allow to build a “scalar field” and draw contour lines of it: useful for fog of wars, liquid effects, paint the ground, etc…

### Draw pathfinding

Draw the pathfinding of an object using a shape painter.

# UI

### Checkbox

Checkbox that can be toggled by a left-click or touch.

### Draggable slider

A draggable slider that users can move to select a numerical value.

### Flexbox

Allows using flexbox for positioning objects.

### Align object on the screen

Align objects on the scene relatively to the window (or screen size).

### Cursor type

Provides an action to change the type of the cursor, and a behavior to change the cursor when an object is hovered.

### Language

Get the preferred language of the user, set on their browser or device.

### Toggle switch

Toggle switch that users can click or touch.

### Star Rating

A rating system indicated by stars using the shape painter.

### Iframe

Create or delete an iframe to embed websites.

### Input Validation

Conditions and expressions to check, sanitize and manipulate strings.

### Pop-up

Display pop-ups to alert, ask confirmation, and let user type a response in text box.

### Mobile virtual keyboard for Text Entry

Allow to open and close the virtual keyboard on mobile (iOS, Android). To be used with Text Entry objects.

### Clipboard

Read and write the clipboard.

### Emojis

Display emoji characters in text objects and store them in strings.

### Console (commands handler for Text Entry objects)

A console behavior that allows parsing simple commands inputed via a Text Entry object.

# Inputs

### Back button

Adds interactions with the back button.

### Cursor Movement Conditions

Conditions checking the cursor movement (is it moving, is it still since a specified duration…).

### Double-click

Check for a double-click with a mouse, or a double-tap on a touchscreen.

### Gamepads (controllers)

Add support for gamepads (or other controllers) to your game, giving access to information such as button presses, axis positions, trigger pressure, etc…

### Mouse Helper

Check if a sprite has been clicked and turn any objects into a cursor.

### Mouse Pointer Lock

This behavior removes the limit on the distance the mouse can move and hides the cursor.

### Behavior Remapper

Quickly remap Behavior controls to different keys.

### Konami Code

Allows to input the classic Konami Code (“Up, Up, Down, Down, Left, Right, Left, Right, B, A”) into a scene for cheats and easter eggs.

### Swipe Detector

A behavior that detects a swipe being done on the screen, and that gives you information about it, including the direction and the length.

# Math and technical tools

### Base conversion

Provides conversion expressions for numbers in different bases.

### Color Conversion

Expressions to convert color values between various formats (RGB, HSV, HSL, named colors) and get colors between 2 others.

### Random Color Generator

Create a random color for a scene, an object, or any other color input.

### Read pixels

Read the values of pixels on the screen.

### Choose a random value

Choose a random value in a list of strings or numbers.

### Extended math support

Additional math functions and constants as expressions and conditions.

### Compressor

Compress and decompress strings.

### Hash

Hash with MD5 or SHA256.

### Unique Identifiers

A collection of UID generation expressions.

### Unicode

Provides conversion tools for Ascii and Unicode characters.

### Time formatting

Expressions to transform time in seconds to format like HH:MM:SS. Ideal to display timers on screen.

### Regular Expressions

Functions for using regular expressions to manipulate strings.

### Rotate a string 13 characters

This extension rotates all alphabetic characters in a string by 13 characters.

### Array tools

A collection of utilities and tools for working with arrays.

### Extended variables support

Add conditions, actions and expressions to check for the existence of a variable, copy variables, delete existing ones from memory, and create dynamic variables.

### Current game version

Get the current game version from GDevelop game properties.

### FPS

Adds expressions and a behavior to get and display the game FPS.

# Network

### Discord rich presence (Windows, Mac, Linux)

Adds discord rich presence to your games.

### Mobile In-App Purchase (experimental)

Add items to buy directly in your game (“In-App Purchase”), for games published on Android or iOS.

### Yandex.Games SDK

An extension for Yandex.Games SDK that lets you easily integrate games created by GDevelop into the Yandex.Games platform.

### Webpage URL tools (Web browser)

Allows to read URL on Web browser exports of the game and manipulate URLs generally.

### Internet Connectivity

Checks if the device running the game is connected to the internet.

### MQTT Client (advanced)

An MQTT client for GDevelop: allow connections to a MQTT server and send/receive messages.

### WebSocket client

A WebSocket client for fast client-server networking.

### Advanced p2p event handling

Allows handling all received P2P events at once instead of one per frame. It is more complex but also potentially more performant.

**Comment by 4ian** [comment]
I would see these categories presented like in the asset store :) 
But instead of an image, a big icon and the category name.

<img width="877" alt="image" src="https://user-images.githubusercontent.com/1280130/180844283-b39c1652-f9b3-4978-b9e1-64d7b455e774.png">
